src/lib/group.ts

```diff
@@ -0,0 +1,364 @@
+import { Scalar, withMessage } from './core.js';
+import { Field, FieldVar, isField } from './field.js';
+import { Bool, Snarky } from '../snarky.js';
+import { Field as Fp } from '../provable/field-bigint.js';
+import { Pallas } from '../bindings/crypto/elliptic_curve.js';
+
+export { Group };
+
+type FieldLike = FieldVar | Field | number | string | bigint;
+type GroupLike = {
+  x: FieldLike;
+  y: FieldLike;
+};
+
+/**
+ * An element of a Group.
+ */
+class Group {
+  x: Field;
+  y: Field;
+
+  /**
+   * The generator `g` of the Group.
+   */
+  static get generator() {
+    return new Group({ x: Pallas.one.x, y: Pallas.one.y });
+  }
+
+  /**
+   * Coerces anything group-like to a {@link Group}.
+   */
+  constructor({
+    x,
+    y,
+  }: {
+    x: FieldVar | Field | number | string | bigint;
+    y: FieldVar | Field | number | string | bigint;
+  }) {
+    this.x = isField(x) ? x : new Field(x);
+    this.y = isField(y) ? y : new Field(y);
+
+    if (this.#isConstant()) {
+      const { add, mul, square } = Fp;
+
+      let x_bigint = this.x.toBigInt();
+      let y_bigint = this.y.toBigInt();
+
+      let onCurve =
+        add(mul(x_bigint, mul(x_bigint, x_bigint)), Pallas.b) ===
+        square(y_bigint);
```

Comment:
I notice there's no way to pass in a `zero` / `infinity` element here. however, not all methods enforce this restriction in the same way. for example, you can to `g.add(g.neg())` and will end up with the group element `(1, 1)` which supposedly stands for `zero`

I think the whole class should pick one consistent way of treating `zero`. It should be consistent between in-snark and out of snark, and could be one of the following:
1. we never ever represent `zero` as a Group. methods that would result in `zero` will throw
2. we have one unique way to represent `zero`. something like `(0, 0)` would make sense.

it's ok if you can use `zero` in some but not all operations on group elements. however, if the constructor and `Group.check()` doesn't allow `zero`, then I'd argue no method should allow you to create `zero`. In order to preserve constraints, I think this might be what you have to do

Reply:
That's a good catch - I'll have to take another look at this.
OCaml returns `(0, 0)` for constants, whereas `(12664759760331458874453076485325239921471337210849432813230171084403110838275, 19449452489080454700052938888178047022259553573804486106032048451047634501628)` for variables

Reply:
although allowing certain functions to handle the `zero` element would require less constraints, I feel like we are limiting certain low-level operations that may require a `zero` element for certain use case, wdyt?

Reply:
Gregor and I had a quick huddle, we thought about leaving it as simple as possible but making it uniform for now. I am about to move the entire 'add` algorithm to JS as well to gain a better understanding, because right now, the OCaml implementation also returns something "weird" for the `zero element (see comment above), We also had the idea of using a `Safe` and `Unsafe` namespace, similar to `Bool`

Reply:
safe and unsafe is a great idea 👍 , but yeah can come after